lxd/auth/oidc/oidc.go

```diff
@@ -292,81 +337,164 @@ func (o *Verifier) setAccessTokenVerifier() error {
-		}
+func (*Verifier) setCookies(w http.ResponseWriter, secureCookie *securecookie.SecureCookie, sessionID uuid.UUID, idToken string, refreshToken string, delete bool) error {
+	idTokenCookie := http.Cookie{
+		Name:     cookieNameIDToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteStrictMode,
+	}
 
-		http.SetCookie(w, &idTokenCookie)
+	refreshTokenCookie := http.Cookie{
+		Name:     cookieNameRefreshToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteStrictMode,
 	}
 
-	if refreshToken != "" || delete {
-		refreshTokenCookie := http.Cookie{
-			Name:     cookieNameRefreshToken,
-			Path:     "/",
-			Secure:   true,
-			HttpOnly: true,
-			SameSite: http.SameSiteStrictMode,
-		}
+	sessionIDCookie := http.Cookie{
+		Name:     cookieNameSessionID,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteStrictMode,
+	}
 
-		if delete {
-			refreshTokenCookie.Value = ""
-			refreshTokenCookie.Expires = time.Unix(0, 0)
-		} else {
-			refreshTokenCookie.Value = refreshToken
-		}
+	if delete {
+		idTokenCookie.Expires = time.Unix(0, 0)
+		refreshTokenCookie.Expires = time.Unix(0, 0)
+		sessionIDCookie.Expires = time.Unix(0, 0)
 
+		http.SetCookie(w, &idTokenCookie)
 		http.SetCookie(w, &refreshTokenCookie)
+		http.SetCookie(w, &sessionIDCookie)
+		return nil
 	}
 
+	encodedIDTokenCookie, err := secureCookie.Encode(cookieNameIDToken, idToken)
+	if err != nil {
+		return fmt.Errorf("Failed to encrypt ID token: %w", err)
+	}
+
+	encodedRefreshToken, err := secureCookie.Encode(cookieNameRefreshToken, refreshToken)
+	if err != nil {
+		return fmt.Errorf("Failed to encrypt refresh token: %w", err)
+	}
+
+	sessionIDCookie.Value = sessionID.String()
+	idTokenCookie.Value = encodedIDTokenCookie
+	refreshTokenCookie.Value = encodedRefreshToken
+
+	http.SetCookie(w, &idTokenCookie)
+	http.SetCookie(w, &refreshTokenCookie)
+	http.SetCookie(w, &sessionIDCookie)
 	return nil
 }
 
-// NewVerifier returns a Verifier.
-func NewVerifier(issuer string, clientid string, audience string) (*Verifier, error) {
-	cookieKey, err := uuid.New().MarshalBinary()
+func (o *Verifier) secureCookieFromSession(sessionID uuid.UUID) (*securecookie.SecureCookie, error) {
+	salt, err := sessionID.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to generate a session ID: %w", err)
+	}
+
+	clusterPrivateKey := o.clusterCert.PrivateKey()
+	keyDerivationFunc := hkdf.New(sha512.New, clusterPrivateKey, salt, []byte("ENCRYPTION"))
+	cookieHashKey := make([]byte, 64)
+	_, err = io.ReadFull(keyDerivationFunc, cookieHashKey)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to create UUID: %w", err)
+		return nil, fmt.Errorf("Failed creating secure cookie hash key: %w", err)
 	}
 
-	verifier := &Verifier{issuer: issuer, clientID: clientid, audience: audience, cookieKey: cookieKey}
+	keyDerivationFunc = hkdf.New(sha512.New, clusterPrivateKey, salt, []byte("INTEGRITY"))
```

Comment:
As an optimisation here, you could directly use the `hkdf.Extract` and `hkdf.Expand` APIs (which is what `hkdf.New` is doing anyway) like this:

```
prk := hkdf.Extract(sha512.New, clusterPrivateKey, salt)
keyDerivationFunc := hkdf.Expand(sha512.New, prk, []byte("ENCRYPTION")
cookieBlockKey := make([]byte, 32)
if _, err = io.ReadFull(keyDerivationFunc, cookieBlockKey); err != nil {
    return nil, fmt.Errorf("Failed creating secure cookie block key: %w", err)
}
keyDerivationFunc = hkdf.Expand(sha512.New, prk, []byte("INTEGRITY")
cookieHashKey := make([]byte, 64)
if _, err = io.ReadFull(keyDerivationFunc, cookieBlockKey); err != nil {
    return nil, fmt.Errorf("Failed creating secure cookie hash key: %w", err)
}
```

Reply:
Thanks, I've made this change.